### PR TITLE
Better cross compilation support

### PR DESCRIPTION
**Pull request by messense**
Originally I want to use [autotools-rs](https://github.com/lu-zero/autotools-rs), but it doesn't support Rust 1.36.0.

This enables building with [cargo-zigbuild](https://github.com/messense/cargo-zigbuild):

```bash
cargo zigbuild --target x86_64-unknown-linux-gnu
```

### CODE REVIEW

File: libffi-sys-rs/build/not_msvc.rs [hunk 1]
```diff
@@ -61,6 +63,53 @@ pub fn configure_libffi(prefix: PathBuf, build_dir: &Path) {
...
+    command.env("CXX", cxx_compiler.path());
+
+    for &(ref k, ref v) in c_compiler.env().iter() {
+        command.env(k, v);
+    }
+
+    for &(ref k, ref v) in cxx_compiler.env().iter() {
+        command.env(k, v);
+    }
+
```
Comment (YorickPeterse) [inline libffi-sys-rs/build/not_msvc.rs:112]: Please run `rustfmt` on this file if you haven't already done so, to ensure the formatting is consistent.

File: libffi-sys-rs/Cargo.toml [hunk 2]
```diff
@@ -20,3 +20,6 @@ features = ["system"]
 
 [target.'cfg(target_env = "msvc")'.build-dependencies]
 cc = "1.0.48"
+
+[build-dependencies]
+cc = "1.0.73"
```
Comment (YorickPeterse) [inline libffi-sys-rs/Cargo.toml:25]: 1. Patch versions aren't needed, instead `cc = "^1.0"` should be good enough
2. If we always require `cc` as a build dependency, we don't need it as a conditional dependency above

**Review by YorickPeterse — CHANGES_REQUESTED** [review]

File: libffi-sys-rs/build/not_msvc.rs [hunk 3]
```diff
@@ -49,10 +49,12 @@ pub fn configure_libffi(prefix: PathBuf, build_dir: &Path) {
     command
         .arg("configure")
         .arg("--with-pic")
+        .arg("--disable-shared")
```
[the remark this answers is not in the document]
Comment (messense) [reply libffi-sys-rs/build/not_msvc.rs:52]: We are going to link the static version so it's a waste of time building a shared library.

https://github.com/tov/libffi-rs/blob/1e7c30376d3477b6727de265fa67725bdfcc22b6/libffi-sys-rs/build/not_msvc.rs#L37

File: libffi-sys-rs/build/not_msvc.rs (hunk 1 above)
Comment (messense) [reply libffi-sys-rs/build/not_msvc.rs:89]: I've added a rustfmt check job to CI.

File: libffi-sys-rs/build/not_msvc.rs [hunk 4]
```diff
@@ -61,6 +63,53 @@ pub fn configure_libffi(prefix: PathBuf, build_dir: &Path) {
...
     }
 
+    let mut c_cfg = cc::Build::new();
+    c_cfg
+        .cargo_metadata(false)
+        .target(&target)
+        .warnings(false)
+        .host(&host);
+    let mut cxx_cfg = cc::Build::new();
+    cxx_cfg
```
[the remark this answers is not in the document]
Comment (messense) [reply libffi-sys-rs/build/not_msvc.rs]: It's not needed indeed, it was copied from autotools-rs.

**Comment by YorickPeterse** [comment]
Windows GNU builds are failing with the following error:

```
  = note: D:\a\libffi-rs\libffi-rs\target\debug\build\libffi-sys-5c1d0e050ef2f003\out\libffi-root\lib\libffi.a(closures.o): In function `internal_malloc_stats':
          D:\a\libffi-rs\libffi-rs\target\debug\build\libffi-sys-5c1d0e050ef2f003\out\libffi-build\x86_64-pc-windows-gnu/../src/dlmalloc.c:2951: undefined reference to `__imp___acrt_iob_func'
          D:\a\libffi-rs\libffi-rs\target\debug\build\libffi-sys-5c1d0e050ef2f003\out\libffi-build\x86_64-pc-windows-gnu/../src/dlmalloc.c:2952: undefined reference to `__imp___acrt_iob_func'
          D:\a\libffi-rs\libffi-rs\target\debug\build\libffi-sys-5c1d0e050ef2f003\out\libffi-build\x86_64-pc-windows-gnu/../src/dlmalloc.c:2953: undefined reference to `__imp___acrt_iob_func'
```

I'm guessing some compile options may be missing?

**Comment by messense** [comment]
x86_64-pc-windows-gnu test only fails on Rust 1.36.0, kinda odd, I'm not sure why.

**Comment by YorickPeterse** [comment]
@messense Past builds/PRs worked just fine though, so this is the result of something that the `cc` crate is doing. Since Rust 1.36 is ancient at this point I'm fine with bumping the minimum version (e.g. to 1.48 as Debian stable seems to still be using that), but not if newer versions of Rust are also failing. Could you check to see how newer versions are doing? Thanks!

**Comment by messense** [comment]
Test passed on Rust 1.48.0. 

https://github.com/messense/libffi-rs/runs/5441820451?check_suite_focus=true

**Comment by messense** [comment]
https://github.com/rust-lang/rust/pull/67429/ might be related, it was released in Rust 1.43.0.

**Review by YorickPeterse — APPROVED** [review]

**Comment by YorickPeterse** [comment]
@messense Thanks! This has been released in libffi-sys 2.0 and libffi-rs 3.0 :tada:

**Pull request merged**